core/src/main/scala/akka/persistence/r2dbc/journal/JournalDao.scala

```diff
@@ -242,58 +249,99 @@ private[r2dbc] class JournalDao(journalSettings: R2dbcSettings, connectionFactor
+        connection =>
+          connection
+            .createStatement(selectLowestSequenceNrSql)
+            .bind(0, persistenceId),
+        row => {
+          val seqNr = row.get(0, classOf[java.lang.Long])
+          if (seqNr eq null) 0L else seqNr.longValue
+        })
+      .map(r => if (r.isEmpty) 0L else r.head)(ExecutionContexts.parasitic)
+
+    if (log.isDebugEnabled)
+      result.foreach(seqNr => log.debug("Lowest sequence nr for persistenceId [{}]: [{}]", persistenceId, seqNr))
+
+    result
+  }
+
+  private def highestSeqNr(persistenceId: String, toSequenceNr: Long): Future[Long] = {
+    if (toSequenceNr == Long.MaxValue)
+      readHighestSequenceNr(persistenceId, 0L)
+    else
+      Future.successful(toSequenceNr)
+  }
+
+  def deleteEventsTo(persistenceId: String, toSequenceNr: Long, resetSequenceNumber: Boolean): Future[Unit] = {
+
+    def deleteBatch(from: Long, to: Long): Future[Unit] = {
+      r2dbcExecutor
+        .updateOne(s"delete [$persistenceId]") { connection =>
           connection
             .createStatement(deleteEventsSql)
             .bind(0, persistenceId)
-            .bind(1, toSequenceNr)
+            .bind(1, from)
+            .bind(2, to)
         }
-      } else {
+        .map(deletedRows =>
+          if (log.isDebugEnabled) {
+            log.debug(
+              "Deleted [{}] events for persistenceId [{}], from seq num [{}] to [{}]",
+              deletedRows,
+              persistenceId,
+              from,
+              to)
+          })(ExecutionContexts.parasitic)
+    }
 
-        val entityType = PersistenceId.extractEntityType(persistenceId)
-        val slice = persistenceExt.sliceForPersistenceId(persistenceId)
-
-        val deleteMarkerSeqNrFut =
-          if (toSequenceNr == Long.MaxValue)
-            readHighestSequenceNr(persistenceId, 0L)
-          else
-            Future.successful(toSequenceNr)
-
-        deleteMarkerSeqNrFut.flatMap { deleteMarkerSeqNr =>
-          def bindDeleteMarker(stmt: Statement): Statement = {
-            stmt
-              .bind(0, slice)
-              .bind(1, entityType)
-              .bind(2, persistenceId)
-              .bind(3, deleteMarkerSeqNr)
-              .bind(4, "")
-              .bind(5, "")
-              .bind(6, 0)
-              .bind(7, "")
-              .bind(8, Array.emptyByteArray)
-              .bind(9, true)
-          }
-
-          r2dbcExecutor
-            .update(s"delete [$persistenceId]") { connection =>
-              Vector(
-                connection
-                  .createStatement(deleteEventsSql)
-                  .bind(0, persistenceId)
-                  .bind(1, toSequenceNr),
-                bindDeleteMarker(connection.createStatement(insertDeleteMarkerSql)))
-            }
-            .map(_.head)
-        }
+    val batchSize = journalSettings.cleanupSettings.eventsJournalDeleteBatchSize
+
+    def deleteInBatches(from: Long, maxTo: Long): Future[Unit] = {
```

Comment:
Add some tests to EventSourcedCleanupSpec. Configure the test with a smaller batch size than default, but larger than what is used for the existing tests. Write more events and then verify that things are deleted in similar ways as existing tests:
* "delete events for one persistenceId"
* "delete events for one persistenceId, but keep seqNr"

If you also want to verify that it batches as expected I agree that it would be easiest to do that with the `LoggingTestKit`.

Reply:
added, not very beautiful ones, but at least I covered the new functionality